// lib/common.js
var socket=require("./socket")
var mysql=require("./mysql")
var func=require("./func")
var async=require("async")

//获取房间信息
exports.getRoomInfo=function(RoomCode,cb)
{
    try
    {
        mysql.query("select * From RoomInfo where RoomCode=" + RoomCode,function(err,Row)
        {
            cb (err,Row);
        });
    }
    catch (e)
    {
        func.log("Err:getRoomInfo "+e)
    }
}
//获取用户信息
exports.getUserInfo=function (UserCode,cb)
{
    try
    {
        mysql.query("select * from RoomUser where UserCode="+UserCode,function(err,Row)
        {
            cb(err,Row);
        });
    }
    catch (e)
    {
        func.log("Err:getUserInfo "+e)
    }
}
//获取用户金币
exports.getUserMoney=function (UserCode,PassWD,cb)
{
    try
    {
        var SQL="select UserMoney,UserType from UserInfo where UserCode="+UserCode;
        if (PassWD)
            SQL+=" and PassWD='"+PassWD+"'"
        mysql.query(SQL,function(err,Row)
        {
            if (cb)
                if (Row)
                    cb(err,Row.UserMoney,Row.UserType);
                else
                    cb(err);
        });
    }
    catch (e)
    {
        func.log("Err:getUserMoney "+e)
    }
}
//保存用户金币
exports.setUserMoney=function (UserCode,Money,cb)
{
    try
    {
        mysql.exec("update UserInfo set  UserMoney ="+Money+" where UserCode="+UserCode,function(err,result)
        {
            if (cb)
                cb(err,result);
        });
    }
    catch (e)
    {
        func.log("Err:setUserMoney "+e)
    }
}
//保存交易记录(unc)
exports.saveUncMoneyRecord=function (RecordType,RecordName,SenderCode,SendMoney,SenderLess,cb)
{
    try
    {
        var SQL = "Insert into MoneyRecord(RecordType,SenderCode,SendMoney,DealTime,SenderLast)";
        SQL += "VALUES("+RecordType+","+SenderCode+","+SendMoney+",'"+func.now()+"',"+SenderLess+")";

        mysql.exec(SQL,function(err,result)
            {

                if (cb)
                    cb(err,result)
            }
        );
    }
    catch (e)
    {
        func.log("Err:saveMoneyRecord "+e)
    }
}
//保存交易记录
exports.saveMoneyRecord=function (RecordType,RecordName,SenderCode,SendMoney,SenderLess,cb)
{
    try
    {
        var SQL = "Insert into UserMoney(RecordType,SenderCode,SendMoney,RecordName,DealTime,SenderLess)";
        SQL += "VALUES("+RecordType+","+SenderCode+","+SendMoney+",'"+RecordName+"','"+func.now()+"',"+SenderLess+")";

        mysql.exec(SQL,function(err,result)
            {
                if (!err)
                    socket.send(SenderCode,"Money",SendMoney,SenderLess);
                if (cb)
                    cb(err,result)
            }
        );
    }
    catch (e)
    {
        func.log("Err:saveMoneyRecord "+e)
    }
}
//删除用户
exports.userQuit=function (UserCode,OnResult)
{
    try
    {

        var RoomCode;
        async.series([GetRoomCode, DeleteOrder, DeleteUser, SendQuit], function (err, result) {
            if (OnResult)
                OnResult(err);
        });
        function GetRoomCode(cb)
        {
            mysql.query("select RoomCode From RoomUser where UserCode=" + UserCode,function(err,Row)
            {

                if (Row)
                    RoomCode=Row.RoomCode;
                cb (err);

            });
        }
        function DeleteOrder(cb) {

            //删除麦序
            mysql.exec("delete From RoomOrder where UserCode=" + UserCode,function(err,result){cb(err)})
        }
        //删除用户
        function DeleteUser(cb)
        {

            mysql.exec("delete From RoomUser where UserCode=" + UserCode,function(err,result){cb(err)})
        }
        //广播用户退出
        function SendQuit(cb) {
            saveServerMsg(RoomCode,0, "UserQuit", UserCode);
            cb()
        }
    }
    catch (e)
    {
        func.log("Err:userQuit "+e)
    }

}
//保存广播
exports.saveServerMsg=saveServerMsg;
function saveServerMsg (ToRoomCode,ToUserCode,Action)
{
    try
    {
        var Arr=[];
        for (var i = 3;i < arguments.length;i++)
            Arr.push(arguments[i]);
        var Data=JSON.stringify(Arr);
        Data='{"action":"'+Action+'","data":'+Data+'}';
        mysql.list("select ServerIP,ServerPort from SysServer where ServerType=1",function(err,Row) {
            for (var i in Row)
                mysql.exec("insert into RoomMsg set Msg='"+Data+"',RoomCode="+ToRoomCode+",UserCode="+ToUserCode+",ServerIP='" + Row[i].ServerIP + "',ServerPort=" + Row[i].ServerPort);
        });
    }
    catch (e)
    {
        func.log("Err:saveServerMsg "+e)
    }
}
exports.makeHornHead=function (RoomCode,UserCode,NickName,Face)
{
    try
    {
        var Str = '<a href=&quot;event:Room' + RoomCode + '&quot;><font color=&quot;#ff0000&quot;><b><u>(' + RoomCode + ')</u></b></font></a>';
        Str+='<FONT SIZE=&quot;16&quot; LETTERSPACING=&quot;8&quot;><img width=&quot;16&quot; height=&quot;16&quot; src=&quot;' + Face + '&quot;/></font>';
        Str += '<a href=&quot;event:User' + UserCode + '&quot;><font color=&quot;#006699&quot;><b><u>' + NickName + '[' + UserCode + ']</u></b></font></a>：';
        return Str;
    }
    catch (e)
    {
        func.log("Err:makeHornHead "+e)
    }
}
exports.sendRichGift=function (client)
{
    mysql.query("select * from RoomGift ", function (err, row) {
        if (client)
            client.send("RichGift",row)
        else
            mysql.list("select UserCode from RoomUser",function (err, rows)
            {
                for (var i in rows)
                {
                    var userCode=rows[i].UserCode;
                    socket.send(userCode,"RichGift",row)
                }
            });
    })
}
//创建发言到空间
exports.saveToZone=function (UserCode,Text,RoomCode)
{
    try
    {

        SQL="insert into ZoneChat set UserCode="+UserCode+",Text='"+Text+"',RoomCode="+RoomCode+",Type=1,Time="+func.time();
        mysql.exec(SQL);
    }
    catch (e)
    {
        func.log("Err:SaveToZone "+e)
    }
}
