// lib/mongo.js
var  mongodb = require('mongodb');
var  func = require('lib/func');
var server;
var  db;
var Conn;
var collectionList={}
exports.init = function(host,port,userName,password,dbName,cb){
    server  = new mongodb.Server(host, port, {auto_reconnect:true});
    db = new mongodb.Db(dbName, server, {safe:true});
    db.open(function(err, db) {
        if (err)
            func.Log("MogoDB连接出错:"+err)
        else
            console.log("MogoDB连接成功。")
        cb(err)
    });
};
function getCollection(collectionName,cb)
{

    db.collection(collectionName, function(err, collection){

        if(err)
            func.Log("MogoDB获取出错:"+collectionName+"\n"+err)

        cb(err,collection)
    });
}
//查询单条数据
exports.query=function(collectionName,data,cb) {
    getCollection(collectionName,function(err,collection)
    {
        if(err)
            return cb(err)
        try
        {
            collection.findOne(data,cb);
        }
        catch (error)
        {
            func.log("数据结果处理出错:" + error+"\n"+collectionName+"\n"+data)
        }

    })
}

//查询多条数据
exports.list=function(collectionName,data,cb) {
    getCollection(collectionName,function(err,collection)
    {
        if(err)
            return cb(err)
        try
        {
            collection.find().toArray(cb);
        }
        catch (error)
        {
            func.log("数据结果处理出错:" + error+"\n"+collectionName+"\n"+data)
        }
    })
}
//修改数据
exports.update=function(collectionName,where,data,multi,cb) {
    getCollection(collectionName,function(err,collection)
    {
        if(err)
            return cb(err)

        collection.update(where,{$set:data},{multi:multi},function(err){
            if (cb)
                cb(err)
        });
    })
}
//插入数据
exports.insert=function(collectionName,data,cb) {
    if (!collectionList[collectionName]) {
        getCollection(collectionName, function (err, collection) {
            if (err)
                return cb(err)
            collectionList[collectionName]=collection;
            console.log(11)
            collection.save(data, function (err) {
                if (cb)
                    cb(err)
            });
        })
    }
    else
    {
        collectionList[collectionName].save(data, function (err) {
            if (cb)
                cb(err)
        });
    }
}
//删除数据
exports.delete=function(collectionName,data,cb) {
    getCollection(collectionName,function(err,collection)
    {
        if(err)
            return cb(err)
        collection.remove(data,function(err){
            if (cb)
                cb(err)
        });
    })
}
